sopel/modules/emailcheck.py

```diff
@@ -0,0 +1,282 @@
+    alert(bot, f"WARNING: User {nick} attempted to register a nick with suspicious domain {email.domain}.")
+
+def fetch_IPQS_email_score(
+    email_addr: str,
+    key: str,
+    fast: bool = True
+    ) -> tuple[float, bool, bool]: #score, disposable, has recent abuse flag set
+    '''Perform lookup on a specific email adress using ipqualityscore.com'''
+    email_str = urllib.parse.quote(email_addr)
+    faststr = str(bool(fast)).lower() #lower + handle None and other garbage
+    params = urllib.parse.urlencode({'fast': faststr})
+    with urllib.request.urlopen(
+        f"https://ipqualityscore.com/api/json/email/{key}/{email_str}?{params}") as url:
+            data = json.loads(url.read().decode())
+            LOGGER.debug(data)
+    if not data['success']:
+        errstr = f"{email_addr} lookup failed with {data['message']}"
+        LOGGER.error(errstr)
+        raise RuntimeError(errstr)
+    return (data['fraud_score'], data["disposable"], data["recent_abuse"])
+
+def get_email_score_from_db(session, email):
+    query_result = session.query(KnownEmails)\
+        .filter(KnownEmails.domain == email.domain)\
+        .one_or_none()
+    if query_result:
+        #Any known problematic provider should've been BADMAILed by now, but...
+        return DomainInfo(query_result.score,
+                          query_result.flag_disposable,
+                          query_result.flag_recent_abuse)
+
+def store_email_score_in_db(session, email, nick, IPQSresult):
+    new_known_email = KnownEmails(domain= email.doman[:DOMAIN_LEN],
+                                    first_nick= nick,
+                                    score= IPQSresult[0],
+                                    flag_disposable= IPQSresult[1],
+                                    flag_recent_abuse= IPQSresult[2])
+    session.add(new_known_email)
+    session.commit()
+
+def retrieve_score(bot, email, nick):
+    session = bot.db.ssession()
+    try:
+        if retval := get_email_score_from_db(session, email):
+            return retval
+        else:
+            if IPQSresult := fetch_IPQS_email_score(email, config.emailcheck.IPQS_key):
+                store_email_score_in_db(session, email, nick, IPQSresult)
+                return IPQSresult
+            else: #Shouldn't be possible
+                raise RuntimeError("Couldn't retrieve IPQS!")
+    except SQLAlchemyError:
+        session.rollback()
+        raise
+    finally:
+        session.remove()
+
+def check_email(bot, email, nick):
+    if any(map(email.endswith, DEFAULT_EXEMPT_SUFFIXES)):
+        #email is exempt
+        LOGGER.info(f'Email {email} used by {nick} is on the exemption list.')
+        return None # No lookup, no result
+    #Check database
+    else:
+        return retrieve_score(bot, email, nick)
+
+# <NickServ> ExampleAccount REGISTER: ExampleNick to foo@example.com
+# (note the 0x02 bold chars)
+@module.rule(r'(\S*)\s*REGISTER: ?([\S]+?)? to ?(\S+)@(\S+?)?$')
```

Comment:
I assume the bold chars are pasted directly into the file - I think it would be more readable to specify them explictly in the regex with `\u0002`

Reply:
will do